Bio/bgzf.py

```diff
@@ -247,19 +247,56 @@
 binary mode, and decode the appropriate fragments yourself.
 """
 
+from __future__ import annotations
+
+import typing as tp
+
 import struct
 import sys
 import zlib
 
 from builtins import open as _open
 
+if tp.TYPE_CHECKING:
+    import _typeshed
+    import io
+    import collections.abc as cx_abc
+    from typing_extensions import TypeAlias, TypeGuard
+
+_BufferT = tp.TypeVar("_BufferT", str, bytes)
+_Self = tp.TypeVar("_Self")
+# fmt: off
+# Writing and updating modes
+_ReadingBinaryMode: TypeAlias = tp.Literal["rb", "br"]
+_ReadingTextMode: TypeAlias = tp.Literal["r", "tr", "rt"]
+_WritingAppendingBinaryMode: TypeAlias = tp.Literal["wb", "bw", "ab", "ba", "wb+", "w+b", "+wb", "bw+", "b+w", "+bw", "ab+", "a+b", "+ab", "ba+", "b+a", "+ba"]
+_WritingAppendingTextMode: TypeAlias = tp.Literal["w", "wt", "tw", "a", "at", "ta", "w+", "+w", "wt+", "w+t", "+wt", "tw+", "t+w", "+tw", "a+", "+a", "at+", "a+t", "+at", "ta+", "t+a", "+ta"]
+_reading_binary_modes: tp.Final[tuple[_ReadingBinaryMode, ...]] = tp.get_args(_ReadingBinaryMode)
+_reading_text_modes: tp.Final[tuple[_ReadingTextMode, ...]] = tp.get_args(_ReadingTextMode)
+_writing_appending_binary_modes: tp.Final[tuple[_WritingAppendingBinaryMode, ...]] = tp.get_args(_WritingAppendingBinaryMode)
+_writing_appending_text_modes: tp.Final[tuple[_WritingAppendingTextMode, ...]] = tp.get_args(_WritingAppendingTextMode)
+# fmt: on
+
+
 _bgzf_magic = b"\x1f\x8b\x08\x04"
 _bgzf_header = b"\x1f\x8b\x08\x04\x00\x00\x00\x00\x00\xff\x06\x00\x42\x43\x02\x00"
 _bgzf_eof = b"\x1f\x8b\x08\x04\x00\x00\x00\x00\x00\xff\x06\x00BC\x02\x00\x1b\x00\x03\x00\x00\x00\x00\x00\x00\x00\x00\x00"
 _bytes_BC = b"BC"
 
 
-def open(filename, mode="rb"):
+# fmt: off
+@tp.overload
+def open(filename: _typeshed.StrPath, mode: _ReadingBinaryMode = "rb") -> BgzfReader[bytes]: ...
```

Comment:
Is it these lines triggering ``D103 missing docstring in public functions``? That feels like a false positive - indeed https://github.com/PyCQA/pydocstyle/issues/419 agreed.

However, pydocstyle is now deprecated. I will look at droping this from our flake8 configuration and enabling it via ruff as suggested.

Reply:
Yep - `@typing.overload`ed functions are usually considered fully-fledged functions, so docstring style-checkers may complain if you don't add something to them. I don't use flake8 in my own projects, but I had a similar experience with https://github.com/HunterMcGushion/docstr_coverage.

I don't know if I'd consider it strictly a false positive, just because if you have more than 3 overloads that have wildly different signatures, it's probably not conducive to cram all of their explanations into one function's docstring.